### PR TITLE
lib: posix: Switch to use zephyr_interface_library_named cmake directive

### LINKED ISSUE #15627: Application compiled with CONFIG_POSIX_API doesn't have access to POSIX headers
Turns out, when an application (e.g. a sample) has an option CONFIG_POSIX_API=y in its prj.conf, it doesn't have include/posix/ path added to its compiler include path. This is completely unexpected given the https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/posix/CMakeLists.txt#L6 directive. Turns out that a few POSIX tests we have work that around by explicitly adding that include path explicitly in app's CMakeLists.txt: https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/posix/common/CMakeLists.txt#L7

This is completely not adequate behavior, requiring obscure options to use Zephyr POSIX subsystem, and leaking and repeating details of internal setup into each and every POSIX app.

### PR DESCRIPTION
**Pull request by pfalcon**
Similar to how other sub-libraries are defined in Zephyr tree, e.g.
"fs", "lgvl", etc. This is supposed to help with the need to
explicitly add posix include path to each and every application using
POSIX subsys.

Fixes: #15627

### CODE REVIEW

**Comment by SebastianBoe** [comment]
We are also going to need a APP_LINK_WITH_PTHREAD Kconfig option as shown here:

https://github.com/zephyrproject-rtos/zephyr/commit/4c60c510c3350818b675d8ecc99dd1e46843321d

This is to allow the user to opt-out of this feature.

**Comment by pfalcon** [comment]
These changes are for #15627, and currently don't work as expected (and thus, should not be merged). Discussion of the issue with these changes are at: https://github.com/zephyrproject-rtos/zephyr/issues/15627#issuecomment-489643141

**Comment by pfalcon** [comment]
> We are also going to need a APP_LINK_WITH_PTHREAD Kconfig option as shown here:

So, current issue is that it simply doesn't work as expected. So, let's leave discussion of opt-out for later. (It would start with the fact that there's opt-in with CONFIG_POSIX_API in the first place.)

**Comment by SebastianBoe** [comment]
Sorry, I meant the design-decision for why this option pattern exists is to allow user to opt-out of the feature.

The feature itself does require the option to be present for it to at all work.

This is documented in the usage documentation of the function here:

https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/extensions.cmake#L485

I see now that the wrong word is used in the documentation, it says "should", but should say "must".

**Comment by pfalcon** [comment]
> The feature itself does require the option to be present for it to at all work.

Thanks, that helped, but I now see include order conflicts between libc vs POSIX headers (before they were in right order, now they're in wrong). Investigating...

**Comment by pfalcon** [comment]
@PiotrZierhoffer, @tgorochowik: This is an example of trying to improve a little bit build process of POSIX apps.

File: tests/posix/common/CMakeLists.txt
Comment (SebastianBoe) [inline tests/posix/common/CMakeLists.txt]: Code should not be commented out.

Comment (pfalcon) [reply tests/posix/common/CMakeLists.txt]: Fixed, and also removed this directive from the other posix test.

**Review by SebastianBoe — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
Sorry for taking a while to review, can you rebase and resolve the conflicts.

**Review by tyrantlizard77 — APPROVED** [review]

File: lib/posix/Kconfig
Comment (nashif) [inline lib/posix/Kconfig]: should be APP_LINK_WITH_POSIX

**Review by nashif — DISMISSED** [review]

File: lib/posix/Kconfig
Comment (pfalcon) [reply lib/posix/Kconfig]: > should be APP_LINK_WITH_POSIX

No, as explained by @SebastianBoe, this should match the existing config symbols, that's why it's APP_LINK_WITH_PTHREAD.

Comment (nashif) [reply lib/posix/Kconfig]: which existing config symbols? If anything is still using PTHREAD, change it POSIX as well.

Comment (pfalcon) [reply lib/posix/Kconfig]: > If anything is still using PTHREAD, change it POSIX as well.

But this is not subject of this change. The subject of this change is "lib: posix: Switch to use zephyr_interface_library_named cmake directive".

Exploration of more deep changes to POSIX subsys was done in e.g. https://github.com/zephyrproject-rtos/zephyr/pull/12984 . Feel free to merge it if you like it.

Comment (nashif) [reply lib/posix/Kconfig]: > But this is not subject of this change. The subject of this change is "lib: posix: Switch to use zephyr_interface_library_named cmake directive".

it becomes a subject of this change when you introduce a new Kconfig using the old name and we will be stuck with it.

Comment (pfalcon) [reply lib/posix/Kconfig]: > we will be stuck with it.

Why would we be stuck with? We'll change it (and hundreds of other mistakes and bugs in Zephyr) when it comes to it.

All in all, I'll defer to @galak, who assigned me to work on POSIX subsys improvements, regarding whether that should be part of this PR or not. In the meantime, working on other assignment from him, and can't work further on this, beyond keeping the build green.

**Comment by pfalcon** [comment]
@galak 

> can you rebase and resolve the conflicts.

Done, build is green.

**Comment by pfalcon** [comment]
Rebased on the latest master.

**Comment by pfalcon** [comment]
> should be APP_LINK_WITH_POSIX

Ok, made this PR start with a patch to rename CMake library PTHREAD -> posix_subsys.

Rebased on master, CI passes. @nashif, as your feedback is addressed, and I'm not sure if you may be on vacation or away, dismissing your review requesting the rename.

**Comment by pfalcon** [comment]
@carlescufi, @galak, @ioannisg: Please consider merging this long-waiting, long-approved PR.

**Comment by ioannisg** [comment]
> 
> 
> @carlescufi, @galak, @ioannisg: Please consider merging this long-waiting, long-approved PR.

Thanks. No rush on this, anyways (it's a bug fix and can be done after feature freeze)

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
